File: products/compliance/src/analyzers/config-checker.ts
import type { Analyzer, Finding, RuleDefinition, ScannedFile } from '../types.js';
import { readFileSync, existsSync } from 'node:fs';
import { dirname, join, basename } from 'node:path';

/**
 * Analyzes project configuration files for security and compliance issues.
 *
 * Checks include:
 * - .env files not excluded by .gitignore
 * - Dockerfile running as root or missing HEALTHCHECK
 * - Missing CI/CD configuration
 * - Wildcard CORS origins in source files
 */
export class ConfigChecker implements Analyzer {
  name = 'config-checker';

  async analyze(files: ScannedFile[], rules: RuleDefinition[]): Promise<Finding[]> {
    const myRules = rules.filter((r) => r.analyzer === this.name);
    if (myRules.length === 0) return [];

    // Build lookup from pattern -> rules (multiple rules may share a pattern)
    const rulesByPattern = new Map<string, RuleDefinition[]>();
    for (const rule of myRules) {
      const existing = rulesByPattern.get(rule.pattern) ?? [];
      existing.push(rule);
      rulesByPattern.set(rule.pattern, existing);
    }

    const findings: Finding[] = [];

    // ── .env not in .gitignore ──────────────────────────────────────
    this.checkEnvNotGitignored(files, rulesByPattern, findings);

    // ── package.json checks ─────────────────────────────────────────
    this.checkPackageJson(files, rulesByPattern, findings);

    // ── Dockerfile checks ───────────────────────────────────────────
    this.checkDockerfile(files, rulesByPattern, findings);

    // ── Wildcard CORS ───────────────────────────────────────────────
    this.checkWildcardCors(files, rulesByPattern, findings);

    return findings;
  }

  /**
   * If a .env file exists in scanned files, check that .gitignore in the
   * same directory contains a .env entry.
   */
  private checkEnvNotGitignored(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const envFiles = files.filter((f) => basename(f.path) === '.env');
    for (const envFile of envFiles) {
      const dir = dirname(envFile.path);
      const gitignorePath = join(dir, '.gitignore');
      let gitignored = false;
      try {
        const content = readFileSync(gitignorePath, 'utf-8');
        const lines = content.split('\n').map((l) => l.trim());
        gitignored = lines.some(
          (line) => line === '.env' || line === '.env*' || line === '*.env',
        );
      } catch {
        // No .gitignore — .env is not gitignored
      }
      if (!gitignored) {
        this.addFindings(findings, 'env-not-gitignored', rulesByPattern, {
          file: envFile.relativePath,
          evidence: '.env not listed in .gitignore',
        });
      }
    }
  }

  /**
   * Parse package.json and check for:
   * - Unpinned dependencies (^ or ~ prefixes)
   * - Missing engines field
   * - Missing lockfile (package-lock.json)
   * - Missing CI configuration
   */
  private checkPackageJson(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const packageFiles = files.filter((f) => basename(f.path) === 'package.json');
    for (const pkgFile of packageFiles) {
      let pkg: Record<string, unknown>;
      try {
        const content = readFileSync(pkgFile.path, 'utf-8');
        pkg = JSON.parse(content) as Record<string, unknown>;
      } catch {
        continue;
      }

      // Unpinned dependencies
      const deps = pkg.dependencies as Record<string, string> | undefined;
      if (deps) {
        const unpinned = Object.entries(deps).filter(
          ([, version]) => version.startsWith('^') || version.startsWith('~'),
        );
        if (unpinned.length > 0) {
          this.addFindings(findings, 'unpinned-deps', rulesByPattern, {
            file: pkgFile.relativePath,
            evidence: `Unpinned: ${unpinned.map(([n]) => n).join(', ')}`,
          });
        }
      }

      // Missing engines
      if (!pkg.engines) {
        this.addFindings(findings, 'missing-engines', rulesByPattern, {
          file: pkgFile.relativePath,
          evidence: 'No engines field in package.json',
        });
      }

      // Missing lockfile
      const lockfilePath = join(dirname(pkgFile.path), 'package-lock.json');
      if (!existsSync(lockfilePath)) {
        this.addFindings(findings, 'missing-lockfile', rulesByPattern, {
          file: pkgFile.relativePath,
          evidence: 'No package-lock.json found',
        });
      }

      // No CI config
      const hasCi = files.some(
        (f) =>
          f.relativePath.startsWith('.github/workflows') ||
          basename(f.path) === '.gitlab-ci.yml' ||
          basename(f.path) === 'Jenkinsfile',
      );
      if (!hasCi) {
        this.addFindings(findings, 'no-ci-config', rulesByPattern, {
          file: pkgFile.relativePath,
          evidence: 'No CI/CD configuration found',
        });
      }
    }
  }

  /**
   * Check Dockerfile for:
   * - Running as root (USER root or no USER directive)
   * - Missing HEALTHCHECK directive
   */
  private checkDockerfile(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const dockerfiles = files.filter((f) => basename(f.path) === 'Dockerfile');
    for (const dockerfile of dockerfiles) {
      let content: string;
      try {
        content = readFileSync(dockerfile.path, 'utf-8');
      } catch {
        continue;
      }

      // Check for root user
      const lines = content.split('\n');
      const userDirectives = lines.filter((line) => /^\s*USER\s+/i.test(line));
      const isRoot =
        userDirectives.length === 0 ||
        userDirectives.some((line) => /^\s*USER\s+root\s*$/i.test(line));

      if (isRoot) {
        // Find the line number of USER root, or report on last line
        let lineNum: number | undefined;
        for (let i = 0; i < lines.length; i++) {
          if (/^\s*USER\s+root\s*$/i.test(lines[i])) {
            lineNum = i + 1;
            break;
          }
        }
        this.addFindings(findings, 'docker-root-user', rulesByPattern, {
          file: dockerfile.relativePath,
          line: lineNum,
          evidence: userDirectives.length === 0 ? 'No USER directive' : 'USER root',
        });
      }

      // Check for HEALTHCHECK
      const hasHealthcheck = lines.some((line) => /^\s*HEALTHCHECK\s+/i.test(line));
      if (!hasHealthcheck) {
        this.addFindings(findings, 'docker-no-healthcheck', rulesByPattern, {
          file: dockerfile.relativePath,
          evidence: 'No HEALTHCHECK directive in Dockerfile',
        });
      }
    }
  }

  /**
   * Check source files for wildcard CORS configuration.
   */
  private checkWildcardCors(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const TEXT_EXTENSIONS = new Set([
      'ts', 'js', 'py', 'go', 'java', 'rb',
      'yaml', 'yml', 'json', 'toml',
      'env', 'cfg', 'conf', 'ini', 'xml', 'properties',
    ]);

    for (const file of files) {
      if (!TEXT_EXTENSIONS.has(file.extension.toLowerCase())) continue;
      if (file.size > 500 * 1024) continue;

      let content: string;
      try {
        content = readFileSync(file.path, 'utf-8');
      } catch {
        continue;
      }

      if (content.includes('\0')) continue;

      const lines = content.split('\n');
      for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (
          line.includes('Access-Control-Allow-Origin: *') ||
          /origin:\s*['"]\*['"]/.test(line)
        ) {
          this.addFindings(findings, 'wildcard-cors', rulesByPattern, {
            file: file.relativePath,
            line: i + 1,
            evidence: 'Wildcard CORS origin detected',
          });
          break; // One finding per file is sufficient
        }
      }
    }
  }

  /**
   * Create findings for all rules matching a given pattern.
   * A single pattern may map to rules in multiple frameworks (soc2, hipaa, gdpr).
   */
  private addFindings(
    findings: Finding[],
    pattern: string,
    rulesByPattern: Map<string, RuleDefinition[]>,
    context: { file?: string; line?: number; evidence?: string },
  ): void {
    const matchedRules = rulesByPattern.get(pattern);
    if (!matchedRules || matchedRules.length === 0) return;

    for (const rule of matchedRules) {
      findings.push({
        id: rule.id,
        title: rule.title,
        description: rule.description,
        severity: rule.severity,
        framework: rule.framework,
        controlIds: rule.controls,
        file: context.file,
        line: context.line,
        evidence: context.evidence,
        analyzer: this.name,
        fixable: rule.fixable,
      });
    }
  }
}

File: products/compliance/analyzers/config_checker.go
package analyzers

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/rishav1305/soul/products/compliance/rules"
)

// ConfigChecker implements the Analyzer interface for detecting configuration
// issues: exposed .env files, Docker misconfigurations, missing security
// headers, CORS wildcards, and missing HTTPS redirects.
type ConfigChecker struct{}

// Name returns the analyzer identifier used to match rules.
func (c *ConfigChecker) Name() string {
	return "config-checker"
}

// config checker regex patterns compiled once.
var (
	corsWildcardRegex  = regexp.MustCompile(`(?i)origin\s*:\s*['"]?\*['"]?`)
	dockerLatestRegex  = regexp.MustCompile(`(?i)^FROM\s+\S+:latest`)
	dockerFromRegex    = regexp.MustCompile(`(?i)^FROM\s+(\S+)`)
	dockerUserRegex    = regexp.MustCompile(`(?i)^USER\s+`)
	dockerHealthRegex  = regexp.MustCompile(`(?i)^HEALTHCHECK\s+`)
	httpsRedirectRegex = regexp.MustCompile(`(?i)(https?.*redirect|redirect.*https|force.*ssl|ssl.*redirect|http.*https)`)
)

// Analyze scans files for configuration-level compliance issues.
func (c *ConfigChecker) Analyze(files []ScannedFile, allRules []rules.Rule) ([]Finding, error) {
	myRules := filterRules(allRules, c.Name())
	if len(myRules) == 0 {
		return nil, nil
	}

	rulesByPattern := groupRulesByPattern(myRules)
	var findings []Finding

	// Build a set of file relative paths for quick existence checks.
	fileIndex := make(map[string]ScannedFile)
	for _, f := range files {
		fileIndex[f.RelativePath] = f
	}

	// Check: .env file exists but not in .gitignore
	findings = append(findings, c.checkEnvExposed(files, fileIndex, rulesByPattern)...)

	// Check Dockerfiles
	findings = append(findings, c.checkDockerfiles(files, rulesByPattern)...)

	// Check source files for CORS wildcards
	findings = append(findings, c.checkCORSWildcard(files, rulesByPattern)...)

	// Check for CI/CD configuration
	findings = append(findings, c.checkCIConfig(files, rulesByPattern)...)

	return findings, nil
}

// checkEnvExposed checks if .env exists but is not covered in .gitignore.
func (c *ConfigChecker) checkEnvExposed(files []ScannedFile, fileIndex map[string]ScannedFile, rulesByPattern map[string][]rules.Rule) []Finding {
	var findings []Finding

	// Check if .env file exists
	hasEnv := false
	for _, f := range files {
		base := filepath.Base(f.RelativePath)
		if base == ".env" || strings.HasPrefix(base, ".env.") {
			hasEnv = true
			break
		}
	}
	if !hasEnv {
		return nil
	}

	// Check if .gitignore exists and contains .env
	envGitignored := false
	if gi, ok := fileIndex[".gitignore"]; ok {
		content, err := os.ReadFile(gi.Path)
		if err == nil {
			lines := strings.Split(string(content), "\n")
			for _, line := range lines {
				trimmed := strings.TrimSpace(line)
				if trimmed == ".env" || trimmed == ".env*" || trimmed == ".env.*" {
					envGitignored = true
					break
				}
			}
		}
	}

	if !envGitignored {
		matchedRules := rulesByPattern["env-not-gitignored"]
		for _, rule := range matchedRules {
			findings = append(findings, Finding{
				ID:          rule.ID,
				Title:       rule.Title,
				Description: rule.Description,
				Severity:    rule.Severity,
				Framework:   rule.Framework,
				ControlIDs:  rule.Controls,
				File:        ".env",
				Evidence:    ".env file exists but is not in .gitignore",
				Analyzer:    c.Name(),
				Fixable:     rule.Fixable,
			})
		}
	}

	return findings
}

// checkDockerfiles inspects Dockerfiles for root user and :latest tag usage.
func (c *ConfigChecker) checkDockerfiles(files []ScannedFile, rulesByPattern map[string][]rules.Rule) []Finding {
	var findings []Finding

	for _, file := range files {
		base := filepath.Base(file.RelativePath)
		if !strings.EqualFold(base, "Dockerfile") && !strings.HasPrefix(strings.ToLower(base), "dockerfile.") {
			continue
		}

		content, err := os.ReadFile(file.Path)
		if err != nil {
			continue
		}

		lines := strings.Split(string(content), "\n")
		hasUser := false
		hasHealthcheck := false

		for lineIdx, line := range lines {
			trimmed := strings.TrimSpace(line)

			// Check for USER directive
			if dockerUserRegex.MatchString(trimmed) {
				hasUser = true
			}

			// Check for HEALTHCHECK directive
			if dockerHealthRegex.MatchString(trimmed) {
				hasHealthcheck = true
			}

			// Check for :latest tag in FROM
			if dockerFromRegex.MatchString(trimmed) {
				// Check if it uses :latest or has no tag (implicit latest)
				if dockerLatestRegex.MatchString(trimmed) {
					for _, rule := range rulesByPattern["docker-latest-tag"] {
						findings = append(findings, Finding{
							ID:          rule.ID,
							Title:       rule.Title,
							Description: rule.Description,
							Severity:    rule.Severity,
							Framework:   rule.Framework,
							ControlIDs:  rule.Controls,
							File:        file.RelativePath,
							Line:        lineIdx + 1,
							Evidence:    trimmed,
							Analyzer:    c.Name(),
							Fixable:     rule.Fixable,
						})
					}
				}
			}
		}

		// Missing USER directive → running as root
		if !hasUser {
			for _, rule := range rulesByPattern["docker-root-user"] {
				findings = append(findings, Finding{
					ID:          rule.ID,
					Title:       rule.Title,
					Description: rule.Description,
					Severity:    rule.Severity,
					Framework:   rule.Framework,
					ControlIDs:  rule.Controls,
					File:        file.RelativePath,
					Evidence:    "Dockerfile does not contain a USER directive",
					Analyzer:    c.Name(),
					Fixable:     rule.Fixable,
				})
			}
		}

		// Missing HEALTHCHECK
		if !hasHealthcheck {
			for _, rule := range rulesByPattern["docker-no-healthcheck"] {
				findings = append(findings, Finding{
					ID:          rule.ID,
					Title:       rule.Title,
					Description: rule.Description,
					Severity:    rule.Severity,
					Framework:   rule.Framework,
					ControlIDs:  rule.Controls,
					File:        file.RelativePath,
					Evidence:    "Dockerfile does not contain a HEALTHCHECK directive",
					Analyzer:    c.Name(),
					Fixable:     rule.Fixable,
				})
			}
		}
	}

	return findings
}

// checkCORSWildcard scans source files for CORS wildcard origin patterns.
func (c *ConfigChecker) checkCORSWildcard(files []ScannedFile, rulesByPattern map[string][]rules.Rule) []Finding {
	var findings []Finding

	sourceExts := map[string]bool{
		"ts": true, "js": true, "py": true, "go": true,
		"java": true, "rb": true, "json": true, "yaml": true, "yml": true,
	}

	for _, file := range files {
		if !sourceExts[strings.ToLower(file.Extension)] {
			continue
		}
		if file.Size > maxFileSize {
			continue
		}

		content, err := os.ReadFile(file.Path)
		if err != nil {
			continue
		}

		lines := strings.Split(string(content), "\n")
		for lineIdx, line := range lines {
			if corsWildcardRegex.MatchString(line) {
				for _, rule := range rulesByPattern["wildcard-cors"] {
					findings = append(findings, Finding{
						ID:          rule.ID,
						Title:       rule.Title,
						Description: fmt.Sprintf("%s (found CORS wildcard)", rule.Description),
						Severity:    rule.Severity,
						Framework:   rule.Framework,
						ControlIDs:  rule.Controls,
						File:        file.RelativePath,
						Line:        lineIdx + 1,
						Evidence:    strings.TrimSpace(line),
						Analyzer:    c.Name(),
						Fixable:     rule.Fixable,
					})
				}
			}
		}
	}

	return findings
}

// checkCIConfig checks for the presence of CI/CD configuration files.
func (c *ConfigChecker) checkCIConfig(files []ScannedFile, rulesByPattern map[string][]rules.Rule) []Finding {
	var findings []Finding

	ciPatterns := []string{
		".github/workflows/",
		".gitlab-ci.yml",
		"Jenkinsfile",
		".circleci/",
		".travis.yml",
		"azure-pipelines.yml",
		"bitbucket-pipelines.yml",
	}

	hasCI := false
	for _, file := range files {
		for _, pattern := range ciPatterns {
			if strings.HasPrefix(file.RelativePath, pattern) || file.RelativePath == pattern {
				hasCI = true
				break
			}
		}
		if hasCI {
			break
		}
	}

	if !hasCI {
		for _, rule := range rulesByPattern["no-ci-config"] {
			findings = append(findings, Finding{
				ID:          rule.ID,
				Title:       rule.Title,
				Description: rule.Description,
				Severity:    rule.Severity,
				Framework:   rule.Framework,
				ControlIDs:  rule.Controls,
				Evidence:    "No CI/CD configuration detected",
				Analyzer:    c.Name(),
				Fixable:     rule.Fixable,
			})
		}
	}

	return findings
}
